### DISCUSSION TITLE
ControlNet API parameters not matching documentation

### DISCUSSION THREAD

**Discussion by dayunbao** [Mikubill/sd-webui-controlnet#2265] [Q&A]
I have a question about the `image` or `input_image` and `mask` parameters.  I'm seeing different things in the docs and what is present in the API payload using the extension @Mikubill made.  Looking here https://github.com/Mikubill/sd-webui-controlnet/wiki/API#integrating-sdapiv12img, I see this, which I think needs to be updated:

```
{
  "prompt": "a cinematic shot of an impressive ants war, ant melee, armageddon",
  "sampler_name": "Euler",
  "alwayson_scripts": {
    "controlnet": {
      "args": [
        {
          "input_image": "base64...",
          "model": "diff_control_sd15_depth_fp16 [978ef0a1]"
        }
      ]
    }
  }
}
```

I tried using the web ui, and checking the API payload.  I saw this:
```
"ControlNet": {
            "args": [
                {
                    "batch_images": "",
                    "control_mode": "Balanced",
                    "enabled": true,
                    "guidance_end": 1,
                    "guidance_start": 0,
                    "image": {
                        "image": "long string",
                        "mask": "long string"
                    },
                    "input_mode": "simple",
                    "is_ui": true,
                    "loopback": false,
                    "low_vram": true,
                    "model": "diffusers_xl_canny_full [2b69fca4]",
                    "module": "canny",
                    "output_dir": "",
                    "pixel_perfect": true,
                    "processor_res": 512,
                    "resize_mode": "Crop and Resize",
                    "save_detected_map": true,
                    "threshold_a": 100,
                    "threshold_b": 200,
                    "weight": 1
                },
                {
                    "batch_images": "",
                    "control_mode": "Balanced",
                    "enabled": true,
                    "guidance_end": 1,
                    "guidance_start": 0,
                    "image": {
                        "image": "long string",
                        "mask": "long string",
                    },
                    "input_mode": "simple",
                    "is_ui": true,
                    "loopback": false,
                    "low_vram": true,
                    "model": "kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]",
                    "module": "openpose",
                    "output_dir": "",
                    "pixel_perfect": true,
                    "processor_res": 512,
                    "resize_mode": "Crop and Resize",
                    "save_detected_map": true,
                    "threshold_a": -1,
                    "threshold_b": -1,
                    "weight": 1
                },
```

**Comment by dayunbao** [comment]
This https://github.com/Mikubill/sd-webui-controlnet/wiki/API#controlnetunitrequest-json-object also isn't matching what is shown in the automatically generated docs.

**Comment by dayunbao** [comment]
If I select `pixel_perfect`, do I need to generate my own `mask`?  If so, should I be sending it like this as part of an sdapi/v1/ request:
```
"image": {
    "image": "long string",
    "mask": "long string"
},
```
and like this as a POST detect request:
```
This object describes a ControlNet processing unit entirely. It has the following properties:

"input_image" : image to use in this unit. defaults to null
"mask" : mask pixel_perfect to filter the image. defaults to null
```

**Comment by dayunbao** [comment]
When I use OpenPose with another ControlNet, like Canny edge detection, the OpenPose input image is being ignored.  When I look at the returned images, the OpenPose input image is blank.  So something is going on.  I'm trying to troubleshoot this, and am getting really confused by the different info I'm seeing about what the request body should look like.

**Comment by dayunbao** [comment]
Should I use the test here https://github.com/Mikubill/sd-webui-controlnet/blob/main/tests/web_api/txt2img_test.py as the source of truth for the request body?  I see this request body in the test:

```
 controlnet_unit = {
            "enabled": True,
            "module": "none",
            "model": self.model,
            "weight": 1.0,
            "image": utils.readImage("test/test_files/img2img_basic.png"),
            "mask": utils.readImage("test/test_files/img2img_basic.png"),
            "resize_mode": 1,
            "lowvram": False,
            "processor_res": 64,
            "threshold_a": 64,
            "threshold_b": 64,
            "guidance_start": 0.0,
            "guidance_end": 1.0,
            "control_mode": 0,
            "pixel_perfect": False,
        }
```

**Comment by dayunbao** [comment]
Here some more data.

My request body:
```
{
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "control_mode": 0,
                    "enabled": true,
                    "guidance_end": 1,
                    "guidance_start": 0,
                    "input_image": "long string",
                    "lowvram": true,
                    "model": "diffusers_xl_canny_full [2b69fca4]",
                    "module": "canny",
                    "pixel_perfect": false,
                    "processor_res": 1024,
                    "resize_mode": 1,
                    "threshold_a": 64,
                    "threshold_b": 64,
                    "weight": 1
                },
                {
                    "control_mode": 0,
                    "enabled": true,
                    "guidance_end": 1,
                    "guidance_start": 0,
                    "input_image": "long string",
                    "lowvram": true,
                    "model": "kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]",
                    "module": "openpose",
                    "pixel_perfect": false,
                    "processor_res": 512,
                    "resize_mode": 1,
                    "threshold_a": 64,
                    "threshold_b": 64,
                    "weight": 1
                }
            ]
        }
    },
    "batch_size": 4,
    "cfg_scale": 7,
    "height": 1024,
    "prompt": "A jedi and a sith in a futuristic apartment",
    "sampler_name": "DPM++ 3M SDE Karras",
    "seed": 0,
    "steps": 50,
    "width": 1024
}
```

API logs:
```
Nov 20 08:13:00 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: API 2023-11-20 08:13:00.755472 200 http/1.1 POST /sdapi/v1/txt2img 188.148.78.22 155.215
Nov 20 08:13:01 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: API 2023-11-20 08:13:01.444732 200 http/1.1 GET /sdapi/v1/progress 188.148.78.22 0.0016
Nov 20 08:14:58 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:14:58,205 - ControlNet - INFO - Loading model: diffusers_xl_canny_full [2b69fca4]
Nov 20 08:14:58 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:14:58,287 - ControlNet - INFO - Loaded state_dict from [/home/ec2-user/Code/comic_creator/stable-diffusion-webui/extensions/sd-webui-controlnet/models/diffusers_xl_canny_full.safetensors]
Nov 20 08:14:58 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:14:58,298 - ControlNet - INFO - controlnet_sdxl_config
Nov 20 08:15:00 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: API 2023-11-20 08:15:00.820101 200 http/1.1 GET /sdapi/v1/progress 188.148.78.22 0.0022
Nov 20 08:15:03 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: API 2023-11-20 08:15:03.820489 200 http/1.1 GET /sdapi/v1/progress 188.148.78.22 0.0017
Nov 20 08:15:05 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:05,203 - ControlNet - INFO - ControlNet model diffusers_xl_canny_full [2b69fca4] loaded.
Nov 20 08:15:05 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:05,275 - ControlNet - INFO - Loading preprocessor: canny
Nov 20 08:15:05 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:05,276 - ControlNet - INFO - preprocessor resolution = 1024
Nov 20 08:15:05 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:05,498 - ControlNet - INFO - Loading model: kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb]
Nov 20 08:15:05 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:05,774 - ControlNet - INFO - Loaded state_dict from [/home/ec2-user/Code/comic_creator/stable-diffusion-webui/extensions/sd-webui-controlnet/models/kohya_controllllite_xl_openpose_anime_v2.safetensors]
Nov 20 08:15:06 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:06,101 - ControlNet - INFO - ControlNet model kohya_controllllite_xl_openpose_anime_v2 [b0fa10bb] loaded.
Nov 20 08:15:06 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:06,116 - ControlNet - INFO - Loading preprocessor: openpose
Nov 20 08:15:06 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:06,116 - ControlNet - INFO - preprocessor resolution = 512
Nov 20 08:15:06 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 2023-11-20 08:15:06,162 - ControlNet - INFO - ControlNet Hooked - Time = 8.18048095703125
Nov 20 08:15:06 ip-172-31-42-185.eu-north-1.compute.internal bash[2934]: 0%|          | 0/50 [00:00<?, ?it/s]API 2023-11-20 08:15:06.820442 200 http/1.1 GET /sdapi/v1/progress 188.148.78.22 0.0023

```

Some screenshots:
![response_1](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/4bdbb39c-c9ca-4f11-a481-b43125add057)
![response_2](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/3cd3faa6-e8bd-4319-ba3b-7ac8da74f670)
![workflow](https://github.com/Mikubill/sd-webui-controlnet/assets/11196050/03275342-24e5-4090-8ab5-c4aeff12abc3)

**Comment by huchenlei** [comment]
See https://github.com/Mikubill/sd-webui-controlnet/blob/0efcf8c99345a67b62b6f54683bc046a7627698b/internal_controlnet/external_code.py#L279-L311
for how image is extracted from payload. There are many valid/compatible ways to pass values as input image/mask.

**Comment by dayunbao** [comment]
Awesome.  Thanks.  So I guess I need to provide my own mask if I want to use pixel_perfect?  Can you point me to where the mask is generated so I can figure out how to do it?  Is it here https://github.com/Mikubill/sd-webui-controlnet/blob/aa4747ab5f949b5d7aa1ccae238571ddca6ef33f/scripts/controlnet.py#L142?

**Comment by huchenlei** [comment]
`pixel_perfect` automatically resize your ControlNet input image's dimensions to fit your SD generation target's dimensions. For example, if your generation target is 1024x1024 but your ControlNet input image is 512x512, the ControlNet input image will be automatically scaled up by 2x when pixel perfect is enabled. That is all `pixel_perfect` option does. It does not relate to mask at all.

**Comment by dayunbao** [comment]
Hmm, OK.  A mask is automatically generated in the web UI, but I guess it isn't used for `pixel_perfect`.  This is part of the request body the web UI created:
```
"ControlNet": {
            "args": [
                {
                    "batch_images": "",
                    "control_mode": "Balanced",
                    "enabled": true,
                    "guidance_end": 1,
                    "guidance_start": 0,
                    "image": {
                        "image": "long string",
                        "mask": "long string" <- HERE
                    },
                    "input_mode": "simple",
                    "is_ui": true,
                    "loopback": false,
                    "low_vram": true,
                    "model": "diffusers_xl_canny_full [2b69fca4]",
                    "module": "canny",
                    "output_dir": "",
                    "pixel_perfect": true,
                    "processor_res": 512,
                    "resize_mode": "Crop and Resize",
                    "save_detected_map": true,
                    "threshold_a": 100,
                    "threshold_b": 200,
                    "weight": 1
                },
```

The OpenPose models started magically working again.  I don't know why.  There is definitely some kind of problem, but I don't know what's causing it or why it goes away.  I doubt anyone will spend the time to figure it out (I don't have the time).   Maybe it's something specific to how I'm using it (AWS with Amazon Linux 2)?
